src/front/wgsl/mod.rs

```diff
@@ -856,10 +859,29 @@ impl Parser {
                     }
                     lexer.expect(Token::Paren(')'))?;
                     if components.is_empty() {
-                        crate::Expression::As {
-                            expr: last_component,
-                            kind: kind.ok_or(Error::BadTypeCast(word))?,
-                            convert: true,
+                        let last_component_inner = ctx.resolve_type(last_component)?;
+                        match (inner, last_component_inner) {
+                            (crate::TypeInner::Scalar { .. }, crate::TypeInner::Scalar { .. })
```

Comment:
the type cast should only really work on scalars, vectors, and matrices. We can error on anything else